libs/pilight/protocols/433.92/tfa303221.c

```diff
@@ -0,0 +1,268 @@
+#define ZERO_PULSE 1300
+#define ONE_PULSE 300
+#define AVG_PULSE 800  // limit puls length for binary analyses
+#define FINAL_PULSE_MIN_LENGTH 1000
+#define MIN_RAW_LENGTH 106
+#define MAX_RAW_LENGTH 116
+#define RAW_LENGTH 106
+
+#define HEADER_PULSE_GAP_MIN_LENGTH 800
+#define HEADER_PULSE_COUNT 8  // 4 pairs of high-low pulses
+
+typedef unsigned char uint8_t;
+
+/*
+ * CRC method
+ */
+static uint8_t lfsr_digest8_reflect(uint8_t const message[], int bytes, uint8_t gen, uint8_t key) {
+	uint8_t sum = 0;
+	// Process message from last byte to first byte (reflected)
+	for (int k = bytes - 1; k >= 0; --k) {
+		uint8_t data = message[k];
+		// Process individual bits of each byte (reflected)
+		for (int i = 0; i < 8; ++i) {
+			// fprintf(stderr, "key is %02x\n", key);
+			// XOR key into sum if data bit is set
+			if ((data >> i) & 1) {
+				sum ^= key;
+			}
+
+			// roll the key left (actually the lsb is dropped here)
+			// and apply the gen (needs to include the dropped lsb as msb)
+			if (key & 0x80)
+				key = (key << 1) ^ gen;
+			else
+				key = (key << 1);
+		}
+	}
+	return sum;
+}
+
+static int validate(void) {
+	if (tfa303221->rawlen >= MIN_RAW_LENGTH && tfa303221->rawlen <= MAX_RAW_LENGTH &&
+		tfa303221->raw[tfa303221->rawlen - 2] > FINAL_PULSE_MIN_LENGTH) {
+		return 0;
+	}
+	return -1;
+}
+
+/*
+ * Finds the index of the first payload bit (after the header of 4 pulse/gap pairs)
+ * Returns -1 if no such header was found
+ *
+ * Example pulsetrain
+ * 260 227 505 232 496 478 262 229 497 236 494 483 256 473 295 816 855 821 858 822 855 825 854 233
+ * 500 232 497 481 258 471 259 228 503 475 261 226 503 234 497 477 257 234 499 475 263 224 504 233
+ * 496 237 492 485 253 233 495 483 255 232 496 482 254 233 498 480 256 472 262 226 502 475 257 231
+ * 501 235 502 472 256 476 262 226 501 478 260 227 500 473 259 233 502 472 264 227 502 231 498 479
+ * 263 225 499 233 500 478 258 229 2056 19207
+ */
+static int findPayloadStartIndex(void) {
+	if (tfa303221->rawlen >= MIN_RAW_LENGTH && tfa303221->rawlen <= MAX_RAW_LENGTH) {
+		int headerFound = 0;
+		for (int i = 0; i < tfa303221->rawlen; i++) {
+			// both pulse & gap need to be > HEADER_PULSE_GAP_MIN_LENGTH
+			if (tfa303221->raw[i] >= HEADER_PULSE_GAP_MIN_LENGTH) {
+				headerFound++;
+			} else if (headerFound > 0) {
+				if (headerFound == HEADER_PULSE_COUNT) {
+					return i;  // this is the first pulse/gap pair that didn't look like a header
+				} else {
+					return -headerFound;  // header not yet complete - broken transmission
+				}
+			}
+		}
+	}
+	return -999;
+}
+
+static void parseCode(void) {
+	int payloadStartIndex = findPayloadStartIndex();
+	if (payloadStartIndex > 0) {
+		// convert pulse-train to array of 0/1 int values
+		unsigned int buffer[40];
+		for (int i = 0; i < 40; i++) {	// read 40 value-pairs from raw data
+			int v0 = tfa303221->raw[payloadStartIndex + (i * 2)];
+			int v1 = tfa303221->raw[payloadStartIndex + (i * 2) + 1];
+			if (v0 < 357 && v1 > 357) {
```

Comment:
Isn't this related to the pulse lengths?

Reply:
yes, v0/v1 are pulse length pairs (shot-long or long-short).

Reply:
switched to use of a constant for `357`